core/services/pipeline/runner.go

```diff
@@ -236,133 +301,94 @@ func (r *runner) executeRun(
 		}
 	)
 
-	var taskDAG TaskDAG
-	err := taskDAG.UnmarshalText([]byte(spec.DotDagSource))
+	pipeline, err := Parse([]byte(spec.DotDagSource))
 	if err != nil {
 		return run, nil, false, err
 	}
 
-	headTaskRuns, err := r.memoryTaskRunDAGFromTaskDAG(taskDAG, pipelineInput, txdb)
-	if err != nil {
-		return run, nil, false, err
+	// initialize certain task params
+	txMu := &sync.Mutex{}
+	for _, task := range pipeline.Tasks {
+		if task.Type() == TaskTypeHTTP {
+			task.(*HTTPTask).config = r.config
+		} else if task.Type() == TaskTypeBridge {
+			task.(*BridgeTask).config = r.config
+			task.(*BridgeTask).safeTx = SafeTx{txdb, txMu}
+		}
 	}
 
+	scheduler := newScheduler(pipeline, pipelineInput)
+
 	// TODO: Test with multiple and single null successor IDs
 	// https://www.pivotaltracker.com/story/show/176557536
 
-	// We execute tasks using "fan in" job processing because tasks have only one
-	// output but potentially many inputs. Start one goroutine per "head task",
-	// having it proceed down its chain of successors until:
-	//  - it encounters a task with unfulfilled predecessors
-	//  - it encounters a task that has already been claimed/finished
-	//  - it processes the final task
-
 	var (
-		vars             = NewVarsFrom(map[string]interface{}{"input": pipelineInput})
-		taskRunResults   TaskRunResults
-		taskRunResultsMu sync.Mutex
-		wg               sync.WaitGroup
-		retry            bool
+		vars  = NewVarsFrom(map[string]interface{}{"input": pipelineInput})
+		retry bool
 	)
-	wg.Add(len(headTaskRuns))
-	for _, mtr := range headTaskRuns {
+
+	for taskRun := range scheduler.taskCh {
+		// execute
 		go func(taskRun *memoryTaskRun) {
 			defer func() {
 				if err := recover(); err != nil {
 					logger.Default.Errorw("goroutine panicked executing run", "panic", err, "stacktrace", string(debug.Stack()))
+
 					// No mutex needed: if any goroutine panics, we retry the run.
 					retry = true
```

Comment:
Currently tasks that panic cause the pipeline to retry, but regular errors don't. Should we always retry? Should we even retry on panics, since they're usually unrecoverable?

An idea was to introduce an optional `retry=N` attribute for each task that would tell the scheduler to retry up to N times

Reply:
I don't think we should retry on panics. Error the task.